pkg/volume/util/operationexecutor/operation_executor.go

```diff
@@ -0,0 +1,763 @@
+				volumeToMount.VolumeSpec.Name(),
+				volumeToMount.PodName,
+				volumeToMount.Pod.UID)
+
+			devicePath, err := volumeAttacher.WaitForAttach(
+				volumeToMount.VolumeSpec, waitForAttachTimeout)
+			if err != nil {
+				glog.Errorf(
+					"MountVolume.WaitForAttach failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v",
+					volumeToMount.VolumeName,
+					volumeToMount.VolumeSpec.Name(),
+					volumeToMount.PodName,
+					volumeToMount.Pod.UID,
+					err)
+				return err
+			}
+
+			glog.Infof(
+				"MountVolume.WaitForAttach succeeded for volume %q (spec.Name: %q) pod %q (UID: %q).",
+				volumeToMount.VolumeName,
+				volumeToMount.VolumeSpec.Name(),
+				volumeToMount.PodName,
+				volumeToMount.Pod.UID)
+
+			// Mount device to global mount path
+			err = volumeAttacher.MountDevice(
+				volumeToMount.VolumeSpec,
+				devicePath,
+				volumeAttacher.GetDeviceMountPath(volumeToMount.VolumeSpec))
+			if err != nil {
+				glog.Errorf(
+					"MountVolume.MountDevice failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v",
+					volumeToMount.VolumeName,
+					volumeToMount.VolumeSpec.Name(),
+					volumeToMount.PodName,
+					volumeToMount.Pod.UID,
+					err)
+				return err
+			}
+
+			glog.Infof(
+				"MountVolume.MountDevice succeeded for volume %q (spec.Name: %q) pod %q (UID: %q).",
+				volumeToMount.VolumeName,
+				volumeToMount.VolumeSpec.Name(),
+				volumeToMount.PodName,
+				volumeToMount.Pod.UID)
+
+			// Update actual state of world to refelct volume is globally mounted
+			markDeviceMountedErr := actualStateOfWorld.MarkDeviceAsMounted(
+				volumeToMount.VolumeName)
+			if markDeviceMountedErr != nil {
+				// On failure, just log and exit. The controller will retry
+				glog.Errorf(
+					"MountVolume.MarkDeviceAsMounted failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v",
+					volumeToMount.VolumeName,
+					volumeToMount.VolumeSpec.Name(),
+					volumeToMount.PodName,
+					volumeToMount.Pod.UID,
+					markDeviceMountedErr)
+				return markDeviceMountedErr
+			}
+		}
+
+		// Execute mount
+		mountErr := volumeMounter.SetUp(fsGroup)
+
+		if mountErr != nil {
+			// On failure, just log and exit. The controller will retry
+			glog.Errorf(
+				"MountVolume.SetUp failed for volume %q (spec.Name: %q) pod %q (UID: %q) with: %v",
+				volumeToMount.VolumeName,
+				volumeToMount.VolumeSpec.Name(),
+				volumeToMount.PodName,
+				volumeToMount.Pod.UID,
+				mountErr)
+			return mountErr
+		}
+
+		glog.Infof(
+			"MountVolume.SetUp succeeded for volume %q (spec.Name: %q) pod %q (UID: %q).",
+			volumeToMount.VolumeName,
+			volumeToMount.VolumeSpec.Name(),
+			volumeToMount.PodName,
+			volumeToMount.Pod.UID)
+
+		// Update actual state of world
+		markVolMountedErr := actualStateOfWorld.MarkVolumeAsMounted(
+			volumeToMount.PodName,
+			volumeToMount.Pod.UID,
+			volumeToMount.VolumeName,
+			volumeMounter,
+			volumeToMount.OuterVolumeSpecName)
+		if markVolMountedErr != nil {
```

Comment:
Yes, `MountVolume` will be called again by the reconciler so `WaitForAttach()`, `MountDevice()`, and `Setup()` will all be re-executed, but they should be idempotent so that shouldn't be a problem.